Content.Shared/Implants/SharedSubdermalImplantSystem.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System.Linq;
+using Content.Shared.Actions;
+using Content.Shared.Actions.ActionTypes;
+using Content.Shared.Implants.Components;
+using Content.Shared.Tag;
+using Robust.Shared.Containers;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Implants;
+
+public abstract class SharedSubdermalImplantSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly SharedActionsSystem _actionsSystem = default!;
+    [Dependency] private readonly SharedContainerSystem _container = default!;
+    [Dependency] private readonly TagSystem _tag = default!;
+
+    public const string ImplantSlotId = "implantcontainer";
+    public const string BaseStorageId = "storagebase";
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<SubdermalImplantComponent, EntGotInsertedIntoContainerMessage>(OnInsert);
+        SubscribeLocalEvent<SubdermalImplantComponent, ContainerGettingRemovedAttemptEvent>(OnRemoveAttempt);
+        SubscribeLocalEvent<SubdermalImplantComponent, EntGotRemovedFromContainerMessage>(OnRemove);
+    }
+
+    private void OnInsert(EntityUid uid, SubdermalImplantComponent component, EntGotInsertedIntoContainerMessage args)
+    {
+        if (component.EntityUid == null)
+            return;
+
+        if (component.ImplantAction != null)
+        {
+            var action = new InstantAction(_prototypeManager.Index<InstantActionPrototype>(component.ImplantAction));
+            _actionsSystem.AddAction(component.EntityUid.Value, action, uid);
+        }
+
+        //replace micro bomb with macro bomb
+        if (_container.TryGetContainer(component.EntityUid.Value, ImplantSlotId, out var implantContainer) && _tag.HasTag(uid, "MacroBomb"))
+        {
+            foreach (var implant in implantContainer.ContainedEntities)
+            {
+                if (_tag.HasTag(implant, "MicroBomb"))
+                {
+                    implantContainer.Remove(implant);
+                    QueueDel(implant);
+                }
+            }
+        }
+    }
+
+    private void OnRemoveAttempt(EntityUid uid, SubdermalImplantComponent component, ContainerGettingRemovedAttemptEvent args)
+    {
+        if (component.Permanent && component.EntityUid != null)
+            args.Cancel();
+    }
+
+    private void OnRemove(EntityUid uid, SubdermalImplantComponent component, EntGotRemovedFromContainerMessage args)
+    {
+        if (component.EntityUid == null)
+            return;
+
+        var entCoords = Transform(component.EntityUid.Value).Coordinates;
```

Comment:
Above I made the comment on entityuid and this the kinda thing that makes me suss of lifetime stuff.